plugins/plugin-optimize/plugin.js

```diff
@@ -136,33 +138,36 @@ exports.default = function plugin(config, userDefinedOptions) {
     // optimize based on extension. if it’s not here, leave as-is
     switch (baseExt) {
       case '.css': {
-        // TODO: minify CSS
+        if (options.minifyCSS) {
+          let code = fs.readFileSync(file, 'utf-8');
+          code = minifyCss(code).css;
+          fs.writeFileSync(file, code, 'utf-8');
+        }
         break;
       }
       case '.js':
       case '.mjs': {
         // minify if enabled
         if (options.minifyJS) {
-          try {
-            let code = fs.readFileSync(file, 'utf-8');
-            const minified = await esbuildService.transform(code, {minify: true, target});
-            code = minified.js;
-            fs.writeFileSync(file, code);
-          } catch (err) {
-            console.error(
-              colors.dim('[@snowpack/plugin-optimize]') +
-                `Error minifying JS [${file}]\n${err.toString()}`,
-            );
-          }
+          let code = fs.readFileSync(file, 'utf-8');
+          const minified = await esbuildService.transform(code, {minify: true, target});
+          code = minified.js;
+          fs.writeFileSync(file, code);
         }
         break;
       }
       case '.html': {
+        let code = fs.readFileSync(file, 'utf-8');
         if (options.preloadModules) {
-          let code = preloadModulesInHTML(file, rootDir);
-          // TODO: minify HTML
-          fs.writeFileSync(file, code, 'utf-8');
+          code = preloadModulesInHTML(code, rootDir);
+        }
+        if (options.minifyHTML) {
+          code = minifyHtml(code, {
+            removeComments: true,
+            collapseWhitespace: true,
```

Comment:
This option can have [problematic side effects](http://perfectionkills.com/experimenting-with-html-minifier/#collapse_whitespace) in the absense of the `conservativeCollapse`.  You'll most certainly want to add that option.

Since the idea here is clearly to have an opinionated set of rules, would you be against other non-controversial options that lack side effects? e.g. `sortAttributes` and `sortClassName` which improve gzip compression in larger files

Reply:
yes, since there are no defaults turned on for you I tried to choose a good set.

If you are more familiar with these, I'd love your help changing these defaults to something less error-prone. The idea is that our default behavior should be as minified as possible without introducing any risk of breakage.

A follow-up task will be to let users configure this themselves.

Reply:
Unfortunately a lot of the optimizations can potentially affect a user's script behavior or CSS selectors, such that a user should have to opt-in to them. The only caveat-free ones include:

- `removeComments`
- `collapseWhitespace`, but only in the presence of `conservativeCollapse`
- `decodeEntities` 
- `sortAttributes`

Reply:
> `collapseWhitespace`, but only in the presence of `conservativeCollapse`

+1 with this thinking for sure. Bike-shedding the API a bit, if 2 options can’t co-exist I’m always a fan of enforcing they occupy the same space. e.g.:

```diff
-  collapseWhitespace: true
-  conservativeCollapse: true
+  whitespace: 'COLLAPSE_CONSERVE' // one of 3–4 values you can find by reading the docs
```